0.6 protocol migration: Signed Data Model (#56)
consumer-reports-digital-lab/data-rights-protocol: data-rights-protocol.md

#### 2.04.1 `POST $status_callback` Response

> Privacy Infrastructure Providers SHOULD make a best effort to ensure that a 200 status is recorded for the most recent status update. The body of the callback's response SHOULD be discarded and not be considered for parsing by the Covered Business.

Reviewer: What does it mean for a `200` status to be recorded?  I'm not sure I'm following what's meant here.

| denied      | insuf_verification     | the [in_progress, need_user_verification] stage failed or timed out | processing_details                           | x      |
| denied      | no_match               | CB could not match user identity to data subject                    | processing_details                           | x      |
| denied      | claim_not_covered      | user requesting data not covered under legal bases[2]             | processing_details                           | x      |
| denied      | outside_jurisdiction   | user requesting data under bases they are not covered by[2]       | processing_details                           | x      |
| denied      | too_many_requests      | user has submitted more requests than the CB is legally obliged to process | details?
| denied      | other                  | some other unspecified failure state reached                        | processing_details                           | x      |
| expired     |                        | the time is currently after the `expires_at` in the request.        |                                              | x      |

> [2]: in the case of claim_not_covered, this may be about asking for categories of data which Covered Businesses are not required to present to the User. in the case of outside_jurisdiction, this may be because the business is not honoring CCPA requests for non-California residents and there is no other basis on which to honor the request. [#28](https://github.com/consumer-reports-digital-lab/data-rights-protocol/issues/28) for discussion on `too_many_requests`

Reviewer: ```suggestion
[2]: In the case of claim_not_covered, this may be about asking for categories of data which Covered Businesses are not required to present to the User. In the case of outside_jurisdiction, this may be because the business is not honoring CCPA requests for non-California residents and there is no other basis on which to honor the request. See [#28](https://github.com/consumer-reports-digital-lab/data-rights-protocol/issues/28) for discussion on `too_many_requests`.
```

    
In version 0.6 we continue to use JWT technology but in the future we will be evaluating a move to another cryptographic envelope library called [libsodium](https://doc.libsodium.org/) which eschews the URL-encoding and provides a much less error-prone API **without exchanging any pair-wise secrets**. The Data Rights Protocol organizing group will be responsible for managing a directory mapping `kid` to public API keys for participants in the network.

0.6 should then be seen as a bridge-effort between the current API security paradigm and a final productionized version. Implementers are encouraged to review libsodium and the programming language bindings which are applicable to their technical stack and provide feedback to the Technical Working Group.

We believe that providing these signed messages will ensure message integrity and prevent re-use or re-appropriate of requests: A data rights request represents a single action of one user acting against one covered business one time.

> Privacy Infrastructure Providers should take care to validate the message in this order:

Reviewer: ```suggestion
Privacy Infrastructure Providers MUST validate the message in this order:
```

- That the signature validates to the key associated with the out of band Authorized Agent identity specified in `kid` URL parameter.

> - That the Authorized Agent attribute in the request matches the Authorized Agent identity outside the envelope

Reviewer: ```suggestion
- That the Authorized Agent attribute in the request matches the Authorized Agent identity outside the envelope (this is very important to prevent requests generated by one AA from being reused by another AA)
```

> - That they are the Covered Business specified inside the `aud` claim (this is very important to prevent request reuse)

Reviewer: ```suggestion
- That they are the Covered Business specified inside the `aud` claim (this is very important to prevent requests originally destined for one CB from being resent to other CBs)
```

> - That the current time is after the Timestamp `iat` claim

Reviewer: ```suggestion
- That the current time is after the Timestamp `iat` claim (this is a very important check for clock-skew, to ensure that requests aren't being generated with expiration times far in the future because the clock of the system generating the requests is running fast)
```

> - That the current time is before the Expiration `exp` claim (this is very important to prevent replay attacks)

Reviewer: ```suggestion
- That the current time is before the Expiration `exp` claim (this is very important to prevent old requests from being replayed)
```
